polkadot/runtime/parachains/src/on_demand/mod.rs

```diff
@@ -391,22 +478,35 @@ where
 				Error::<T>::QueueFull
 			);
 
-			// Charge the sending account the spot price. The amount will be teleported to the
-			// broker chain once it requests revenue information.
-			let amt = T::Currency::withdraw(
-				&sender,
-				spot_price,
-				WithdrawReasons::FEE,
-				existence_requirement,
-			)?;
-
-			// Consume the negative imbalance and deposit it into the pallet account. Make sure the
-			// account preserves even without the existential deposit.
-			let pot = Self::account_id();
-			if !System::<T>::account_exists(&pot) {
-				System::<T>::inc_providers(&pot);
+			match payment_type {
+				PaymentType::Balance => {
+					// Charge the sending account the spot price. The amount will be teleported to
+					// the broker chain once it requests revenue information.
+					let amt = T::Currency::withdraw(
+						&sender,
+						spot_price,
+						WithdrawReasons::FEE,
+						existence_requirement,
+					)?;
+
+					// Consume the negative imbalance and deposit it into the pallet account. Make
+					// sure the account preserves even without the existential deposit.
+					let pot = Self::account_id();
+					if !System::<T>::account_exists(&pot) {
+						System::<T>::inc_providers(&pot);
+					}
+					T::Currency::resolve_creating(&pot, amt);
+				},
+				PaymentType::Credits => {
+					// Charge the sending account the spot price in credits.
+					Credits::<T>::try_mutate(&sender, |credits| -> DispatchResult {
+						*credits = credits
+							.checked_sub(&spot_price)
+							.ok_or(Error::<T>::InsufficientCredits)?;
+						Ok(())
+					})?;
```

Comment:
maybe worth deleting the entry if it reaches 0 credits. otherwise we never clean up credits

Reply:
033cb0a2133f76e08afa9cd68b8bffa55e73b832